w1r3/infra/mig/java/main.tf

```diff
@@ -0,0 +1,140 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+variable "project" {
+  type = string
+}
+
+variable "bucket" {
+  type = string
+}
+
+variable "region" {
+  type = string
+}
+
+variable "replicas" {
+  type = number
+}
+
+variable "service_account" {
+  type = string
+}
+
+variable "app_version" {
+  type    = string
+  default = "v1"
+}
+
+locals {
+  # Including the version in the name simplifies updates. An `apply` will
+  # delete the previous version and create a new one.
+  runner-template    = "java-runner-template-${var.app_version}"
+  runner-group       = "java-runner-${var.app_version}"
+  base_instance_name = "w1r3-java"
+  cpus               = 4
+  iterations         = 1000000
+}
+
+resource "google_compute_instance_template" "default" {
+  name = local.runner-template
+  disk {
+    auto_delete  = true
+    boot         = true
+    device_name  = "persistent-disk-0"
+    mode         = "READ_WRITE"
+    source_image = "projects/cos-cloud/global/images/family/cos-stable"
+    type         = "PERSISTENT"
+  }
+  labels = {
+    w1r3-lang = "java"
+  }
+  machine_type = "c2d-standard-${local.cpus}"
+  metadata = {
+    cos-metrics-enabled = true
+    user-data           = <<EOF
+ #cloud-config
+
+users:
+- name: cloudservice
+  uid: 2000
+
+write_files:
+- path: /etc/systemd/system/w1r3@.service
+  permissions: 0644
+  owner: root
+  content: |
+    [Unit]
+    Description=The Java w1r3 continuous benchmark instance %I
+    StartLimitIntervalSec=300
+    StartLimitBurst=3
+
+    [Service]
+    Restart=always
+    RestartSec=1s
+
+    Environment="HOME=/home/cloudservice"
+    ExecStartPre=/usr/bin/docker-credential-gcr configure-docker ; /usr/bin/docker pull gcr.io/${var.project}/w1r3/java:latest
+    ExecStart=/usr/bin/docker run --rm -u 2000 --name=w1r3-java-%i gcr.io/${var.project}/w1r3/java:latest java -jar /r/w1r3.jar -project-id ${var.project} -bucket ${var.bucket} -iterations ${local.iterations} -deployment mig -workers 1
+    ExecStop=/usr/bin/docker stop w1r3-java-%i
+    ExecStopPost=/usr/bin/docker rm w1r3-java-%i
+
+runcmd:
+- docker-credential-gcr configure-docker
+- systemctl daemon-reload
+- for i in $(seq 1 ${local.cpus}); do systemctl start w1r3@$${i}.service; done
```

Comment:
To confirm my understanding of the systemd unit and the cloud config:

The unit will start an individual process with a single worker. Then cloud-config will start multiple services, one for each CPU on the host machine?

Reply:
Yes.  The idea is that measuring CPU / byte is easier if a single thread is doing the uploads and downloads.

Reply:
Makes sense, thanks for confirming